src/build_system.ml

```diff
@@ -544,23 +395,36 @@ module Rule_fn = struct
   let loc () = Fdecl.get loc_decl ()
 end
 
+module Context_or_install = struct
+  type t =
+    | Install of string
+    | Context of string
+
+  let to_sexp = function
+    | Install ctx -> Sexp.List [ Sexp.Atom "install"; Sexp.Atom ctx ]
+    | Context s ->
+      assert (not (s = "install"));
+      Sexp.Atom s
+end
+
+
 type t =
   { (* File specification by targets *)
     files       : Internal_rule.t Path.Table.t
   ; contexts    : Context.t String.Map.t
   ; file_tree   : File_tree.t
   ; dirs : Dir_status.t Path.Table.t
+  ; init_rules : Rules.t Fdecl.t
   ; gen_rules :
-      (dir:Path.t -> string list -> extra_sub_directories_to_keep)
-        String.Map.t Fdecl.t
+      (Context_or_install.t ->
+       (dir:Path.t -> string list -> extra_sub_directories_to_keep) option) Fdecl.t
   ; mutable load_dir_stack : Path.t list
   ; (* Set of directories under _build that have at least one rule and
        all their ancestors. *)
     mutable build_dirs_to_keep : Path.Set.t
-  ; mutable prefix : (unit, unit) Build.t option
   ; hook : hook -> unit
   ; (* Package files are part of *)
-    packages : Package.Name.t Path.Table.t
+    packages : (Path.t -> Package.Name.t list) Fdecl.t
```

Comment:
I think it would be a bit more clear if this function returned a `Package.Name.Set.t` instead.